File: ghidra_scripts/ExportBeaconFlowMetadata.py
# Export BeaconFlow metadata from Ghidra headless mode
#
# Usage:
#   analyzeHeadless <project_dir> <project_name> \
#     -import <binary> \
#     -postScript ExportBeaconFlowMetadata.py <output.json> \
#     -deleteProject
#
# Or on Windows:
#   analyzeHeadless.bat <project_dir> <project_name> \
#     -import <binary> \
#     -postScript ExportBeaconFlowMetadata.py <output.json> \
#     -deleteProject
#
# The output JSON is compatible with BeaconFlow's load_metadata() and
# has the same schema as the IDA export_ida_metadata.py output:
#
#   {
#     "input_path": "...",
#     "image_base": "0x...",
#     "functions": [
#       {
#         "name": "func_name",
#         "start": "0x...",
#         "end": "0x...",
#         "blocks": [
#           { "start": "0x...", "end": "0x...", "succs": ["0x..."] }
#         ]
#       }
#     ]
#   }

from __future__ import print_function

import json
import os
import sys

from ghidra.program.model.block import BasicBlockModel
from ghidra.program.model.address import AddressSet
from ghidra.util.task import ConsoleTaskMonitor


def _hex(value):
    return "0x{:x}".format(value)


def _get_image_base(program):
    return program.getImageBase().getOffset()


def _get_input_path(program):
    path = program.getExecutablePath()
    if path:
        return path
    return program.getName()


def _get_function_name(function, program):
    name = function.getName()
    if name and not name.startswith("FUN_"):
        return name
    return _hex(function.getEntryPoint().getOffset())


def _get_blocks_and_succs(function, block_model, program):
    entry = function.getEntryPoint()
    body = function.getBody()
    monitor = ConsoleTaskMonitor()

    blocks = []
    addr_to_block = {}

    code_blocks = block_model.getCodeBlocksContaining(body, monitor)
    while code_blocks.hasNext():
        cb = code_blocks.next()
        start = cb.getMinAddress().getOffset()
        end = cb.getMaxAddress().getOffset() + 1
        addr_to_block[start] = cb
        blocks.append({
            "start": _hex(start),
            "end": _hex(end),
            "succs": []
        })

    for block_info in blocks:
        start_offset = int(block_info["start"], 16)
        cb = addr_to_block.get(start_offset)
        if cb is None:
            continue
        dest_iter = cb.getDestinations(monitor)
        while dest_iter.hasNext():
            dest = dest_iter.next()
            dest_block = dest.getDestinationBlock()
            succ_start = dest_block.getMinAddress().getOffset()
            block_info["succs"].append(_hex(succ_start))

    return blocks


def export_metadata(output_path):
    program = getCurrentProgram()
    listing = program.getListing()
    fm = program.getFunctionManager()
    block_model = BasicBlockModel(program)
    monitor = ConsoleTaskMonitor()

    functions = []
    func_iter = fm.getFunctions(True)
    for function in func_iter:
        blocks = _get_blocks_and_succs(function, block_model, program)
        functions.append({
            "name": _get_function_name(function, program),
            "start": _hex(function.getEntryPoint().getOffset()),
            "end": _hex(function.getBody().getMaxAddress().getOffset() + 1),
            "blocks": blocks,
        })

    data = {
        "input_path": _get_input_path(program),
        "image_base": _hex(_get_image_base(program)),
        "functions": functions,
    }

    with open(output_path, "w") as f:
        json.dump(data, f, indent=2)

    print("[BeaconFlow] Exported {} functions to {}".format(len(functions), output_path))


if __name__ == "__main__":
    if len(sys.argv) < 2:
        raise SystemExit("usage: ExportBeaconFlowMetadata.py <output.json>")
    export_metadata(sys.argv[1])

File: ghidra_scripts/ExportBeaconFlowMetadata.java
// Export BeaconFlow metadata from Ghidra headless mode
//
// Usage:
//   analyzeHeadless <project_dir> <project_name> \
//     -import <binary> \
//     -postScript ExportBeaconFlowMetadata.java <output.json> \
//     -deleteProject
//
// Output JSON schema (compatible with BeaconFlow load_metadata):
//   {
//     "input_path": "...",
//     "image_base": "0x...",
//     "functions": [
//       {
//         "name": "func_name",
//         "start": "0x...",
//         "end": "0x...",
//         "blocks": [
//           { "start": "0x...", "end": "0x...", "succs": ["0x..."] }
//         ]
//       }
//     ]
//   }

import ghidra.app.script.GhidraScript;
import ghidra.program.model.block.BasicBlockModel;
import ghidra.program.model.block.CodeBlock;
import ghidra.program.model.block.CodeBlockReference;
import ghidra.program.model.listing.Function;
import ghidra.program.model.listing.FunctionManager;
import ghidra.util.task.ConsoleTaskMonitor;

import java.io.FileWriter;
import java.util.LinkedHashMap;
import java.util.ArrayList;

public class ExportBeaconFlowMetadata extends GhidraScript {
    @Override
    public void run() throws Exception {
        String outputPath = getScriptArgs()[0];

        FunctionManager fm = currentProgram.getFunctionManager();
        BasicBlockModel blockModel = new BasicBlockModel(currentProgram);
        ConsoleTaskMonitor monitor = new ConsoleTaskMonitor();

        long imageBase = currentProgram.getImageBase().getOffset();
        String inputPath = currentProgram.getExecutablePath();
        if (inputPath == null) {
            inputPath = currentProgram.getName();
        }

        StringBuilder sb = new StringBuilder();
        sb.append("{\n");
        sb.append("  \"input_path\": \"").append(escapeJson(inputPath)).append("\",\n");
        sb.append("  \"image_base\": \"0x").append(Long.toHexString(imageBase)).append("\",\n");
        sb.append("  \"functions\": [\n");

        Function func = fm.getFunctions(true).next();
        boolean firstFunc = true;
        for (Function f : fm.getFunctions(true)) {
            if (!firstFunc) {
                sb.append(",\n");
            }
            firstFunc = false;

            String name = f.getName();
            if (name == null || name.startsWith("FUN_")) {
                name = "0x" + Long.toHexString(f.getEntryPoint().getOffset());
            }

            long funcStart = f.getEntryPoint().getOffset();
            long funcEnd = f.getBody().getMaxAddress().getOffset() + 1;

            sb.append("    {\n");
            sb.append("      \"name\": \"").append(escapeJson(name)).append("\",\n");
            sb.append("      \"start\": \"0x").append(Long.toHexString(funcStart)).append("\",\n");
            sb.append("      \"end\": \"0x").append(Long.toHexString(funcEnd)).append("\",\n");
            sb.append("      \"blocks\": [\n");

            ArrayList<CodeBlock> blocks = new ArrayList<>();
            var codeBlocks = blockModel.getCodeBlocksContaining(f.getBody(), monitor);
            while (codeBlocks.hasNext()) {
                blocks.add(codeBlocks.next());
            }

            boolean firstBlock = true;
            for (CodeBlock cb : blocks) {
                if (!firstBlock) {
                    sb.append(",\n");
                }
                firstBlock = false;

                long blockStart = cb.getMinAddress().getOffset();
                long blockEnd = cb.getMaxAddress().getOffset() + 1;

                sb.append("        {\"start\": \"0x").append(Long.toHexString(blockStart)).append("\", ");
                sb.append("\"end\": \"0x").append(Long.toHexString(blockEnd)).append("\", ");
                sb.append("\"succs\": [");

                var destIter = cb.getDestinations(monitor);
                boolean firstSucc = true;
                while (destIter.hasNext()) {
                    CodeBlockReference dest = destIter.next();
                    long succStart = dest.getDestinationBlock().getMinAddress().getOffset();
                    if (!firstSucc) {
                        sb.append(", ");
                    }
                    firstSucc = false;
                    sb.append("\"0x").append(Long.toHexString(succStart)).append("\"");
                }

                sb.append("]}");
            }

            sb.append("\n      ]\n");
            sb.append("    }");
        }

        sb.append("\n  ]\n");
        sb.append("}\n");

        try (FileWriter writer = new FileWriter(outputPath)) {
            writer.write(sb.toString());
        }

        println("[BeaconFlow] Exported " + fm.getFunctionCount() + " functions to " + outputPath);
    }

    private String escapeJson(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\")
                .replace("\"", "\\\"")
                .replace("\n", "\\n")
                .replace("\r", "\\r")
                .replace("\t", "\\t");
    }
}
